src/psij/_plugins.py

```diff
@@ -0,0 +1,123 @@
+import importlib
+import inspect
+import logging
+from bisect import bisect_left
+from distutils.versionpredicate import VersionPredicate
+from types import ModuleType
+from typing import Tuple, Dict, List, Union, Type, Any, Optional, TypeVar
+
+from psij._descriptor import _Descriptor, _VersionEntry
+
+logger = logging.getLogger(__name__)
+
+
+def _getfile(o: Optional[Union[Type[Any], ModuleType]]) -> str:
+    if o is None:
+        return ''
+    else:
+        return inspect.getfile(o)
+
+
+def _split_cls_name(cls: str) -> Tuple[str, str]:
+    comps = cls.split('.')
+    return '.'.join(comps[:-1]), comps[-1]
+
+
+T = TypeVar('T')
+
+
+def _register_plugin(desc: _Descriptor, root_path: str, type: str,
+                     store: Dict[str, List[_VersionEntry[T]]]) -> None:
+    """
+    Registers a class with a certain base class through a :class:`~psij._Descriptor`.
+
+    This is used internally to dynamically find and register :class:`~psij.JobExecutor` and
+    :class:`~psij.Launcher` classes.
+
+    Parameters
+    ----------
+    desc:
+        a descriptor whose `cls` property must point to a class that has `cls` as a base class.
+    type:
+        the type of class being registered (either "executor" or "launcher")
+    store:
+        a dictionary where the registered plugins are being stored
+    """
+    module, cls_name = _split_cls_name(desc.cls)
+    cls = None
+    exc = None
+    mod_path = root_path + '/' + module.replace('.', '/') + '.py'
+    try:
+        spec = importlib.util.spec_from_file_location(module, mod_path)
+        if spec is None:
+            raise Exception('Failed to load %s' % mod_path)
+        mod = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(mod)  # type: ignore
+        cls = getattr(mod, cls_name)
+    except Exception as ex:
+        s = str(ex)
+        if 'does_not_exist' not in s:
+            # there is a class used for testing that would print a warning message every time
+            # psij was imported; intercept that message and avoid logging it
```

Comment:
it isn't immediately clear to me what this comment means wrt the surrounding code: you're logging all exceptions from module loading except ones containing the string "does_not_exist" which you are hoping only comes from a test class and not from anything user relevant?

Reply:
could remove this from the core code, and only import the deliberately-broken test case descriptor during test runs? if descriptors can be spread across multiple locations, and there's a specific extra path set already for testing?

Reply:
Fixed by efc7231ddae9375cdb04dd3a1a5a9d2fbcb1a2e2